aptos-move/framework/aptos-stdlib/sources/cryptography/algebra.move

```diff
@@ -42,22 +214,61 @@ module aptos_std::algebra {
         serialize_internal<S, F>(element.handle)
     }
 
-    //
-    // (Public functions end here.)
-    // Native functions begin.
-    //
+    /// Get the order of group `G`, a big integer little-endian encoded as a byte array.
+    public fun order<S>(): vector<u8> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        order_internal<S>()
+    }
 
-    native fun deserialize_internal<G, F>(bytes: &vector<u8>): (bool, u64);
-    native fun add_internal<F>(handle_1: u64, handle_2: u64): u64;
-    native fun serialize_internal<G, F>(handle: u64): vector<u8>;
+    /// Cast an element of a structure `S` to a parent structure `L`.
+    public fun upcast<S,L>(element: &Element<S>): Element<L> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        Element<L> {
+            handle: upcast_internal<S,L>(element.handle)
+        }
+    }
+
+    /// Try casting an element `x` of a structure `L` to a sub-structure `S`.
+    /// Return none if `x` is not a member of `S`.
+    ///
+    /// NOTE: Membership check is performed inside, which can be expensive, depending on the structures `L` and `S`.
+    public fun downcast<L,S>(element_x: &Element<L>): Option<Element<S>> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        let (succ, new_handle) = downcast_internal<L,S>(element_x.handle);
+        if (succ) {
+            some(Element<S> { handle: new_handle })
+        } else {
+            none()
+        }
+    }
+
+    /// Hash an arbitrary-length byte array `msg` into structure `S` using the given `suite`.
+    /// A unique domain separation tag `dst` of size 255 bytes or shorter is required
```

Comment:
Here, I'd like to understand:

1. Not all hashing methods will have a DST, so how will this work for those calls? (e.g., see ristretto255.move hashing natives)
2. Why are we enforcing a max length on the DST? Is this our choice or is this what arkworks does?
   - If it's our choice, we shouldn't be doing it. It's not necessary and might hamper some applications.
   - If it's from arkworks, then this is curve-specific and should not be in the `algebra.move` comments but in `algebra_bls12381.move`.

Reply:
1. If there is `algebra::from_uniform_bytes<S>(bytes: vector<u8>)`, those functions in `ristretto255.move` can fit well into it.
2. Thanks for catching! That can be fixed to support long DSTs.

Reply:
Removed DST limitations and updated gas formula.

Reply:
Great! For hashing methods that do not accept a DST (e.g., hashing in Ristretto255), we will implement the native to abort if a non-empty DST is given.

Reply:
We can later add a `hash_to_no_dst` method too that simply calls `hash_to` with an empty `dst`.